### PR TITLE
added the ability to search book by title and author

### PR DESCRIPTION
**Pull request by odeineka**

### CODE REVIEW

File: src/main/java/com/example/demo/contrtoller/BookController.java [hunk 1]
```diff
@@ -48,4 +50,12 @@ public BookDto updateBook(@PathVariable Long id, @RequestBody CreateBookRequestD
     public void softDeleteBook(@PathVariable Long id) {
         bookService.softDeleteBookById(id);
     }
+
+    @GetMapping("/search")
+    public List<BookDto> search(@RequestParam(required = false) String titlePart,
+                                @RequestParam(required = false) String authorPart) {
+        BookSearchParametersDto searchParameters = new BookSearchParametersDto(
+                titlePart, authorPart);
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/demo/contrtoller/BookController.java:58]:
```suggestion
    public List<BookDto> search(BookSearchParametersDto searchParameters) 
```
you can pass searchParameters directly

File: src/main/java/com/example/demo/dto/BookSearchParametersDto.java [hunk 2]
```diff
@@ -0,0 +1,4 @@
+package com.example.demo.dto;
+
+public record BookSearchParametersDto(String titlePart, String authorPart) {
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/demo/dto/BookSearchParametersDto.java:3]:
```suggestion
public record BookSearchParametersDto(String[] title, String[] author) {
```
better pass arrays for multiple searches

File: src/main/java/com/example/demo/repository/book/spec/AuthorSpecificationProvider.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+import com.example.demo.model.Book;
+import com.example.demo.repository.SpecificationProvider;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@Component
+public class AuthorSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey() {
+        return "author";
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/demo/repository/book/spec/AuthorSpecificationProvider.java:12]: better make it public constant and use it in BookSpecificationBuilder

File: src/main/java/com/example/demo/repository/book/spec/AuthorSpecificationProvider.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    }
+
+    @Override
+    public Specification<Book> getSpecification(String authorPart) {
+        return (root, query, criteriaBuilder) ->
+                criteriaBuilder.like(criteriaBuilder
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/demo/repository/book/spec/AuthorSpecificationProvider.java:18]: if I remember the correct operator LIKE ignore case in MySQL

File: src/main/java/com/example/demo/repository/book/spec/AuthorSpecificationProvider.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+                        .lower(root.get("author")), "%" + authorPart.toLowerCase() + "%");
+    }
+}
+
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/demo/repository/book/spec/AuthorSpecificationProvider.java:22]:
```suggestion
```
only one empty line

File: src/main/java/com/example/demo/repository/book/spec/TitleSpecificationProvider.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
+package com.example.demo.repository.book.spec;
+
+import com.example.demo.model.Book;
+import com.example.demo.repository.SpecificationProvider;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@Component
+public class TitleSpecificationProvider implements SpecificationProvider<Book> {
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/demo/repository/book/spec/TitleSpecificationProvider.java:9]: the same comments

File: src/main/java/com/example/demo/service/impl/BookServiceImpl.java [hunk 7]
```diff
@@ -54,4 +59,13 @@ public void softDeleteBookById(Long id) {
         );
         bookRepository.delete(book);
     }
+
+    @GetMapping("api/books/search")
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/demo/service/impl/BookServiceImpl.java:63]: Why do you need this annotation here?

File: src/main/java/com/example/demo/service/impl/BookServiceImpl.java [hunk 8]
```diff
@@ -54,4 +59,13 @@ public void softDeleteBookById(Long id) {
...
+    public List<BookDto> search(BookSearchParametersDto searchParameters) {
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/demo/service/impl/BookServiceImpl.java:64]: Also, it would be great to add pagination

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s  improve your solution ;)

File: src/main/java/com/example/demo/repository/book/spec/AuthorSpecificationProvider.java (hunk 4 above)
Comment (odeineka) [reply src/main/java/com/example/demo/repository/book/spec/AuthorSpecificationProvider.java:18]: following SOLID principle, to provide durability and exclude some register sensitive cases in future, I' will leave this transfering

File: src/main/java/com/example/demo/contrtoller/BookController.java [hunk 9]
```diff
@@ -13,18 +15,19 @@
...
 import org.springframework.web.bind.annotation.RequestBody;
 import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
 import org.springframework.web.bind.annotation.ResponseStatus;
 import org.springframework.web.bind.annotation.RestController;
 
 @RequiredArgsConstructor
 @RestController
-@RequestMapping("/books")
+@RequestMapping("/api/books")
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/demo/contrtoller/BookController.java:24]: add `server.servlet.context-path=/api` to application.properties

```suggestion
@RequestMapping("/books")
```

File: src/main/java/com/example/demo/contrtoller/BookController.java [hunk 10]
```diff
@@ -48,4 +51,14 @@ public BookDto updateBook(@PathVariable Long id, @RequestBody CreateBookRequestD
     public void softDeleteBook(@PathVariable Long id) {
         bookService.softDeleteBookById(id);
     }
+
+    @GetMapping("/search")
+    public List<BookDto> search(
+            @RequestParam(required = false) String[] titleParts,
+            @RequestParam(required = false) String[] authorParts,
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/demo/contrtoller/BookController.java:58]:
```suggestion
            BookSearchParametersDto searchParameters,
```

File: src/main/java/com/example/demo/repository/SpecificationBuilder.java [hunk 11]
```diff
@@ -0,0 +1,9 @@
+package com.example.demo.repository;
+
+import com.example.demo.dto.BookSearchParametersDto;
+import org.springframework.data.jpa.domain.Specification;
+
+public interface SpecificationBuilder<T> {
+
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/demo/repository/SpecificationBuilder.java:7]: remove empty line
```suggestion
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by JJJazl — APPROVED** [review]

**Pull request merged**